## Preprocessing

In [4]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd
application_df = pd.read_csv("https://static.bc-edx.com/data/dl-1-2/m21/lms/starter/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [5]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
#  YOUR CODE GOES HERE
application_df = application_df.drop(columns=['EIN', 'NAME'])

In [6]:
# Determine the number of unique values in each column.
#  YOUR CODE GOES HERE
unique_values = application_df.nunique()
print(unique_values)

APPLICATION_TYPE            17
AFFILIATION                  6
CLASSIFICATION              71
USE_CASE                     5
ORGANIZATION                 4
STATUS                       2
INCOME_AMT                   9
SPECIAL_CONSIDERATIONS       2
ASK_AMT                   8747
IS_SUCCESSFUL                2
dtype: int64


In [7]:
# Look at APPLICATION_TYPE value counts to identify and replace with "Other"
#  YOUR CODE GOES HERE
# Look at APPLICATION_TYPE value counts
application_type_counts = application_df["APPLICATION_TYPE"].value_counts()

# Display the counts
print(application_type_counts)

APPLICATION_TYPE
T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: count, dtype: int64


In [8]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
#  YOUR CODE GOES HERE
cutoff = 500

# Creating a list of application types to be replaced
application_types_to_replace = application_type_counts[application_type_counts < cutoff].index.tolist()

# Replace in dataframe
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure replacement was successful
application_df['APPLICATION_TYPE'].value_counts()

APPLICATION_TYPE
T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: count, dtype: int64

In [9]:
# Look at CLASSIFICATION value counts to identify and replace with "Other"
#  YOUR CODE GOES HERE
classification_counts = application_df["CLASSIFICATION"].value_counts()

# Displaying the counts
print(classification_counts)

CLASSIFICATION
C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C1248        1
C6100        1
C1820        1
C1900        1
C2150        1
Name: count, Length: 71, dtype: int64


In [10]:
# You may find it helpful to look at CLASSIFICATION value counts >1
#  YOUR CODE GOES HERE
classification_counts_filtered = classification_counts[classification_counts > 1].sort_values(ascending=False)

# Display the counts
print(classification_counts_filtered)

CLASSIFICATION
C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C6000       15
C1800       15
C1250       14
C8200       11
C1238       10
C1278       10
C1237        9
C1235        9
C7210        7
C1720        6
C2400        6
C4100        6
C1257        5
C1600        5
C1260        3
C2710        3
C0           3
C3200        2
C1256        2
C1246        2
C1234        2
C1267        2
Name: count, dtype: int64


In [11]:
# Choose a cutoff value and create a list of classifications to be replaced
cutoff = 1000
classification_counts = application_df['CLASSIFICATION'].value_counts()

# Create the list of classifications to replace based on the cutoff
classifications_to_replace = classification_counts[classification_counts < cutoff].index.tolist()

# Replace in dataframe using the provided loop structure
for cls in classifications_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls, "Other")

# Check to make sure replacement was successful
print(application_df['CLASSIFICATION'].value_counts())

CLASSIFICATION
C1000    17326
C2000     6074
C1200     4837
Other     2261
C3000     1918
C2100     1883
Name: count, dtype: int64


In [12]:
# Convert categorical data to numeric with `pd.get_dummies`
#  YOUR CODE GOES HERE
application_df_dummies = pd.get_dummies(application_df)
application_df_dummies.head()

STATUS  ASK_AMT  IS_SUCCESSFUL  APPLICATION_TYPE_Other  \
0       1     5000              1                   False   
1       1   108590              1                   False   
2       1     5000              0                   False   
3       1     6692              1                   False   
4       1   142590              1                   False   

   APPLICATION_TYPE_T10  APPLICATION_TYPE_T19  APPLICATION_TYPE_T3  \
0                  True                 False                False   
1                 False                 False                 True   
2                 False                 False                False   
3                 False                 False                 True   
4                 False                 False                 True   

   APPLICATION_TYPE_T4  APPLICATION_TYPE_T5  APPLICATION_TYPE_T6  ...  \
0                False                False                False  ...   
1                False                False                False  ...   
2                False                 True                False  ...   
3                False                False                False  ...   
4                False                False                False  ...   

   INCOME_AMT_1-9999  INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  \
0              False                   False                     False   
1               True                   False                     False   
2              False                   False                     False   
3              False                    True                     False   
4              False                   False                      True   

   INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  \
0               False             False                   False   
1               False             False                   False   
2               False             False                   False   
3               False             False                   False   
4               False             False                   False   

   INCOME_AMT_50M+  INCOME_AMT_5M-10M  SPECIAL_CONSIDERATIONS_N  \
0            False              False                      True   
1            False              False                      True   
2            False              False                      True   
3            False              False                      True   
4            False              False                      True   

   SPECIAL_CONSIDERATIONS_Y  
0                     False  
1                     False  
2                     False  
3                     False  
4                     False  

[5 rows x 44 columns]

In [27]:
# Split our preprocessed data into our features and target arrays
#  YOUR CODE GOES HERE

# Defining the target variable 
y = application_df_dummies['IS_SUCCESSFUL']

# Define the feature variables (Dropping more columns)
X = application_df_dummies.drop(
    ['IS_SUCCESSFUL', 'APPLICATION_TYPE_Other', 'INCOME_AMT_1-9999', 'SPECIAL_CONSIDERATIONS_N', 
     'APPLICATION_TYPE_T10', 'INCOME_AMT_50M+'], 
    axis=1
)
# Split the preprocessed data into a training and testing dataset
#  YOUR CODE GOES HERE
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [21]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Compile, Train and Evaluate the Model

In [23]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
#  YOUR CODE GOES HERE
from tensorflow.keras import layers
# Define the model - deep neural net
nn = tf.keras.models.Sequential()

# First hidden layer with 80 units
nn.add(layers.Dense(units=80, input_dim=X_train_scaled.shape[1], activation='relu'))

# Second hidden layer with 30 units
nn.add(layers.Dense(units=30, activation='relu'))

# Output layer with 1 unit
nn.add(layers.Dense(units=1, activation='sigmoid'))

# Check the structure of the model
nn.summary()

Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_3 (Dense)                 │ (None, 80)             │         3,120 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_4 (Dense)                 │ (None, 30)             │         2,430 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_5 (Dense)                 │ (None, 1)              │            31 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 5,581 (21.80 KB)

 Trainable params: 5,581 (21.80 KB)

 Non-trainable params: 0 (0.00 B)

In [24]:
# Compile the model
nn.compile(optimizer='adam', 
           loss='binary_crossentropy', 
           metrics=['accuracy'])

In [25]:
# Train the model
#  YOUR CODE GOES HERE
# Train the model
history = nn.fit(X_train_scaled, 
                 y_train, 
                 epochs=100, 
                 batch_size=len(X_train_scaled), 
                 validation_data=(X_test_scaled, y_test), 
                 verbose=2)

Epoch 1/100
1/1 - 11s - 11s/step - accuracy: 0.4923 - loss: 0.7624 - val_accuracy: 0.5106 - val_loss: 0.7449
Epoch 2/100
1/1 - 0s - 198ms/step - accuracy: 0.5169 - loss: 0.7429 - val_accuracy: 0.5190 - val_loss: 0.7272
Epoch 3/100
1/1 - 0s - 137ms/step - accuracy: 0.5267 - loss: 0.7257 - val_accuracy: 0.5334 - val_loss: 0.7120
Epoch 4/100
1/1 - 0s - 141ms/step - accuracy: 0.5385 - loss: 0.7110 - val_accuracy: 0.5574 - val_loss: 0.6988
Epoch 5/100
1/1 - 0s - 137ms/step - accuracy: 0.5624 - loss: 0.6982 - val_accuracy: 0.5980 - val_loss: 0.6873
Epoch 6/100
1/1 - 0s - 139ms/step - accuracy: 0.6028 - loss: 0.6869 - val_accuracy: 0.6185 - val_loss: 0.6771
Epoch 7/100
1/1 - 0s - 139ms/step - accuracy: 0.6267 - loss: 0.6768 - val_accuracy: 0.6526 - val_loss: 0.6680
Epoch 8/100
1/1 - 0s - 137ms/step - accuracy: 0.6569 - loss: 0.6677 - val_accuracy: 0.6525 - val_loss: 0.6599
Epoch 9/100
1/1 - 0s - 134ms/step - accuracy: 0.6577 - loss: 0.6595 - val_accuracy: 0.6810 - val_loss: 0.6525
Epoch 10/10

In [26]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

215/215 - 0s - 2ms/step - accuracy: 0.7249 - loss: 0.5611
Loss: 0.5611077547073364, Accuracy: 0.7249271273612976


- Despite removing additional columns, the model's accuracy still falls below 75% (0.7249 or 72.49%). In the next step, I'll add more neurons to a hidden layer. 

In [29]:
# Define the model - deep neural net
nn = tf.keras.models.Sequential()

# First hidden layer with 128 units (increased from 80)
nn.add(layers.Dense(units=128, input_dim=X_train_scaled.shape[1], activation='relu'))

# Second hidden layer with 64 units (increased from 30)
nn.add(layers.Dense(units=64, activation='relu'))

# Third hidden layer (newly added) with 32 units
nn.add(layers.Dense(units=32, activation='relu'))

# Output layer with 1 unit
nn.add(layers.Dense(units=1, activation='sigmoid'))

# Check the structure of the model
nn.summary()

Model: "sequential_3"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_10 (Dense)                │ (None, 128)            │         4,992 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_11 (Dense)                │ (None, 64)             │         8,256 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_12 (Dense)                │ (None, 32)             │         2,080 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_13 (Dense)                │ (None, 1)              │            33 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 15,361 (60.00 KB)

 Trainable params: 15,361 (60.00 KB)

 Non-trainable params: 0 (0.00 B)

In [30]:
# Compiling the new model
nn.compile(optimizer='adam', 
           loss='binary_crossentropy', 
           metrics=['accuracy'])

In [31]:
# Training the new model
history = nn.fit(X_train_scaled, 
                 y_train, 
                 epochs=100, 
                 batch_size=len(X_train_scaled), 
                 validation_data=(X_test_scaled, y_test), 
                 verbose=2)

Epoch 1/100
1/1 - 3s - 3s/step - accuracy: 0.5683 - loss: 0.6921 - val_accuracy: 0.6124 - val_loss: 0.6815
Epoch 2/100
1/1 - 0s - 352ms/step - accuracy: 0.6227 - loss: 0.6748 - val_accuracy: 0.6465 - val_loss: 0.6669
Epoch 3/100
1/1 - 0s - 155ms/step - accuracy: 0.6635 - loss: 0.6607 - val_accuracy: 0.6711 - val_loss: 0.6553
Epoch 4/100
1/1 - 0s - 159ms/step - accuracy: 0.6874 - loss: 0.6496 - val_accuracy: 0.6920 - val_loss: 0.6455
Epoch 5/100
1/1 - 0s - 161ms/step - accuracy: 0.7025 - loss: 0.6402 - val_accuracy: 0.6975 - val_loss: 0.6373
Epoch 6/100
1/1 - 0s - 160ms/step - accuracy: 0.7039 - loss: 0.6322 - val_accuracy: 0.6984 - val_loss: 0.6300
Epoch 7/100
1/1 - 0s - 157ms/step - accuracy: 0.7057 - loss: 0.6248 - val_accuracy: 0.6988 - val_loss: 0.6233
Epoch 8/100
1/1 - 0s - 157ms/step - accuracy: 0.7058 - loss: 0.6179 - val_accuracy: 0.7023 - val_loss: 0.6172
Epoch 9/100
1/1 - 0s - 161ms/step - accuracy: 0.7099 - loss: 0.6115 - val_accuracy: 0.6985 - val_loss: 0.6115
Epoch 10/100


In [32]:
# Evaluate the new model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

215/215 - 0s - 2ms/step - accuracy: 0.7270 - loss: 0.5572
Loss: 0.5572186708450317, Accuracy: 0.7269679307937622


- As seen, the model's accuracy increases slightly to 72.70% with the addition of more neurons. In the next step, I add more hidden layers to the code.

In [40]:
from tensorflow.keras import layers
import tensorflow as tf

# Define the model - deep neural net
nn = tf.keras.models.Sequential()

# First hidden layer with 128 units
nn.add(layers.Dense(units=128, input_dim=X_train_scaled.shape[1], activation='relu'))

# Second hidden layer with 64 units
nn.add(layers.Dense(units=64, activation='relu'))

# Third hidden layer with 32 units
nn.add(layers.Dense(units=32, activation='relu'))

# Fourth hidden layer with 16 units (new layer added)
nn.add(layers.Dense(units=16, activation='relu'))

# Fifth hidden layer with 8 units (new layer added)
nn.add(layers.Dense(units=8, activation='relu'))

# Output layer with 1 unit
nn.add(layers.Dense(units=1, activation='sigmoid'))

# Check the structure of the model
nn.summary()

Model: "sequential_6"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_22 (Dense)                │ (None, 128)            │         4,992 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_23 (Dense)                │ (None, 64)             │         8,256 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_24 (Dense)                │ (None, 32)             │         2,080 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_25 (Dense)                │ (None, 16)             │           528 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_26 (Dense)                │ (None, 8)              │           136 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_27 (Dense)                │ (None, 1)              │             9 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 16,001 (62.50 KB)

 Trainable params: 16,001 (62.50 KB)

 Non-trainable params: 0 (0.00 B)

In [41]:
# Compiling the new model
nn.compile(optimizer='adam', 
           loss='binary_crossentropy', 
           metrics=['accuracy'])

In [42]:
# Training the new model
history = nn.fit(X_train_scaled, 
                 y_train, 
                 epochs=100, 
                 batch_size=len(X_train_scaled), 
                 validation_data=(X_test_scaled, y_test), 
                 verbose=2)

Epoch 1/100
1/1 - 10s - 10s/step - accuracy: 0.4493 - loss: 0.6977 - val_accuracy: 0.5566 - val_loss: 0.6934
Epoch 2/100
1/1 - 0s - 259ms/step - accuracy: 0.5623 - loss: 0.6925 - val_accuracy: 0.5614 - val_loss: 0.6898
Epoch 3/100
1/1 - 0s - 307ms/step - accuracy: 0.5681 - loss: 0.6889 - val_accuracy: 0.5603 - val_loss: 0.6864
Epoch 4/100
1/1 - 0s - 169ms/step - accuracy: 0.5703 - loss: 0.6856 - val_accuracy: 0.6146 - val_loss: 0.6834
Epoch 5/100
1/1 - 0s - 172ms/step - accuracy: 0.6216 - loss: 0.6825 - val_accuracy: 0.6191 - val_loss: 0.6799
Epoch 6/100
1/1 - 0s - 360ms/step - accuracy: 0.6260 - loss: 0.6789 - val_accuracy: 0.6497 - val_loss: 0.6763
Epoch 7/100
1/1 - 0s - 393ms/step - accuracy: 0.6568 - loss: 0.6751 - val_accuracy: 0.6571 - val_loss: 0.6727
Epoch 8/100
1/1 - 0s - 167ms/step - accuracy: 0.6630 - loss: 0.6714 - val_accuracy: 0.6777 - val_loss: 0.6687
Epoch 9/100
1/1 - 0s - 192ms/step - accuracy: 0.6856 - loss: 0.6672 - val_accuracy: 0.6583 - val_loss: 0.6644
Epoch 10/10

In [43]:
# Evaluate the new model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

215/215 - 0s - 2ms/step - accuracy: 0.7271 - loss: 0.5574
Loss: 0.557415246963501, Accuracy: 0.7271137237548828


- With the addition of more hidden layers, the model's accuracy increases, albeit slightly, to 72.71%. In the next step, I use different activation functions for the hidden layers.

In [45]:
from tensorflow.keras import layers
import tensorflow as tf

# Define the model - deep neural net
nn = tf.keras.models.Sequential()

# First hidden layer with 128 units and ReLU activation
nn.add(layers.Dense(units=128, input_dim=X_train_scaled.shape[1], activation='relu'))

# Second hidden layer with 64 units and LeakyReLU activation
nn.add(layers.Dense(units=64))
nn.add(layers.LeakyReLU(alpha=0.1))  # alpha defines the slope of the negative part

# Third hidden layer with 32 units and Tanh activation
nn.add(layers.Dense(units=32, activation='tanh'))

# Fourth hidden layer with 16 units and ELU activation
nn.add(layers.Dense(units=16, activation='elu'))

# Fifth hidden layer with 8 units and Sigmoid activation
nn.add(layers.Dense(units=8, activation='sigmoid'))

# Output layer with 1 unit (binary classification)
nn.add(layers.Dense(units=1, activation='sigmoid'))

# Check the structure of the model
nn.summary()

Model: "sequential_8"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_34 (Dense)                │ (None, 128)            │         4,992 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_35 (Dense)                │ (None, 64)             │         8,256 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ leaky_re_lu_1 (LeakyReLU)       │ (None, 64)             │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_36 (Dense)                │ (None, 32)             │         2,080 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_37 (Dense)                │ (None, 16)             │           528 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_38 (Dense)                │ (None, 8)              │           136 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_39 (Dense)                │ (None, 1)              │             9 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 16,001 (62.50 KB)

 Trainable params: 16,001 (62.50 KB)

 Non-trainable params: 0 (0.00 B)

In [46]:
# Compiling the new model
nn.compile(optimizer='adam', 
           loss='binary_crossentropy', 
           metrics=['accuracy'])

In [47]:
# Training the new model
history = nn.fit(X_train_scaled, 
                 y_train, 
                 epochs=100, 
                 batch_size=len(X_train_scaled), 
                 validation_data=(X_test_scaled, y_test), 
                 verbose=2)

Epoch 1/100
1/1 - 3s - 3s/step - accuracy: 0.4680 - loss: 0.7959 - val_accuracy: 0.4659 - val_loss: 0.7865
Epoch 2/100
1/1 - 0s - 167ms/step - accuracy: 0.4680 - loss: 0.7858 - val_accuracy: 0.4659 - val_loss: 0.7770
Epoch 3/100
1/1 - 0s - 163ms/step - accuracy: 0.4680 - loss: 0.7764 - val_accuracy: 0.4659 - val_loss: 0.7681
Epoch 4/100
1/1 - 0s - 166ms/step - accuracy: 0.4680 - loss: 0.7676 - val_accuracy: 0.4659 - val_loss: 0.7598
Epoch 5/100
1/1 - 0s - 163ms/step - accuracy: 0.4680 - loss: 0.7593 - val_accuracy: 0.4659 - val_loss: 0.7521
Epoch 6/100
1/1 - 0s - 165ms/step - accuracy: 0.4680 - loss: 0.7516 - val_accuracy: 0.4659 - val_loss: 0.7450
Epoch 7/100
1/1 - 0s - 164ms/step - accuracy: 0.4680 - loss: 0.7445 - val_accuracy: 0.4659 - val_loss: 0.7384
Epoch 8/100
1/1 - 0s - 167ms/step - accuracy: 0.4680 - loss: 0.7379 - val_accuracy: 0.4659 - val_loss: 0.7323
Epoch 9/100
1/1 - 0s - 165ms/step - accuracy: 0.4680 - loss: 0.7319 - val_accuracy: 0.4659 - val_loss: 0.7267
Epoch 10/100


In [48]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

215/215 - 1s - 5ms/step - accuracy: 0.7270 - loss: 0.5707
Loss: 0.5706889629364014, Accuracy: 0.7269679307937622


- Despite using different activation methods, the model's accuracy still falls below the desired 75%. It drops to 72.70%. In the next step, I add the number of epochs to the training regimen to 200.

In [51]:
from tensorflow.keras import layers
import tensorflow as tf

# Define the model - deep neural net
nn = tf.keras.models.Sequential()

# First hidden layer with 128 units and ReLU activation
nn.add(layers.Dense(units=128, input_dim=X_train_scaled.shape[1], activation='relu'))

# Second hidden layer with 64 units and LeakyReLU activation
nn.add(layers.Dense(units=64))
nn.add(layers.LeakyReLU(alpha=0.1))  # alpha defines the slope of the negative part

# Third hidden layer with 32 units and Tanh activation
nn.add(layers.Dense(units=32, activation='tanh'))

# Fourth hidden layer with 16 units and ELU activation
nn.add(layers.Dense(units=16, activation='elu'))

# Fifth hidden layer with 8 units and Sigmoid activation
nn.add(layers.Dense(units=8, activation='sigmoid'))

# Output layer with 1 unit (binary classification)
nn.add(layers.Dense(units=1, activation='sigmoid'))

# Check the structure of the model
nn.summary()

# Compile the model
nn.compile(optimizer='adam', 
           loss='binary_crossentropy', 
           metrics=['accuracy'])

# Train the model with 100 epochs
history = nn.fit(X_train_scaled, 
                 y_train, 
                 epochs=200,  # Number of epochs
                 batch_size=64, 
                 validation_data=(X_test_scaled, y_test), 
                 verbose=2)

Model: "sequential_10"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_46 (Dense)                │ (None, 128)            │         4,992 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_47 (Dense)                │ (None, 64)             │         8,256 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ leaky_re_lu_3 (LeakyReLU)       │ (None, 64)             │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_48 (Dense)                │ (None, 32)             │         2,080 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_49 (Dense)                │ (None, 16)             │           528 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_50 (Dense)                │ (None, 8)              │           136 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_51 (Dense)                │ (None, 1)              │             9 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 16,001 (62.50 KB)

 Trainable params: 16,001 (62.50 KB)

 Non-trainable params: 0 (0.00 B)

Epoch 1/200
429/429 - 3s - 8ms/step - accuracy: 0.7158 - loss: 0.5796 - val_accuracy: 0.7222 - val_loss: 0.5692
Epoch 2/200
429/429 - 1s - 3ms/step - accuracy: 0.7273 - loss: 0.5597 - val_accuracy: 0.7203 - val_loss: 0.5633
Epoch 3/200
429/429 - 1s - 3ms/step - accuracy: 0.7298 - loss: 0.5552 - val_accuracy: 0.7284 - val_loss: 0.5605
Epoch 4/200
429/429 - 1s - 3ms/step - accuracy: 0.7314 - loss: 0.5528 - val_accuracy: 0.7230 - val_loss: 0.5598
Epoch 5/200
429/429 - 1s - 3ms/step - accuracy: 0.7306 - loss: 0.5515 - val_accuracy: 0.7259 - val_loss: 0.5595
Epoch 6/200
429/429 - 1s - 3ms/step - accuracy: 0.7324 - loss: 0.5497 - val_accuracy: 0.7216 - val_loss: 0.5600
Epoch 7/200
429/429 - 1s - 3ms/step - accuracy: 0.7329 - loss: 0.5486 - val_accuracy: 0.7265 - val_loss: 0.5592
Epoch 8/200
429/429 - 1s - 3ms/step - accuracy: 0.7330 - loss: 0.5475 - val_accuracy: 0.7243 - val_loss: 0.5598
Epoch 9/200
429/429 - 1s - 3ms/step - accuracy: 0.7332 - loss: 0.5469 - val_accuracy: 0.7213 - val_loss:

In [52]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

215/215 - 0s - 2ms/step - accuracy: 0.7245 - loss: 0.5675
Loss: 0.5674553513526917, Accuracy: 0.7244898080825806


- The addition of more epochs to the training regimen decreases the model's accuracy to 72.45%. In the next step, I reduce the number of epochs to the training regimen.

In [56]:
from tensorflow.keras import layers
import tensorflow as tf

# Define the model - deep neural net
nn = tf.keras.models.Sequential()

# First hidden layer with 128 units and ReLU activation
nn.add(layers.Dense(units=128, input_dim=X_train_scaled.shape[1], activation='relu'))

# Second hidden layer with 64 units and LeakyReLU activation
nn.add(layers.Dense(units=64))
nn.add(layers.LeakyReLU(alpha=0.1))  # alpha defines the slope of the negative part

# Third hidden layer with 32 units and Tanh activation
nn.add(layers.Dense(units=32, activation='tanh'))

# Fourth hidden layer with 16 units and ELU activation
nn.add(layers.Dense(units=16, activation='elu'))

# Fifth hidden layer with 8 units and Sigmoid activation
nn.add(layers.Dense(units=8, activation='sigmoid'))

# Output layer with 1 unit (binary classification)
nn.add(layers.Dense(units=1, activation='sigmoid'))

# Check the structure of the model
nn.summary()

# Compile the model
nn.compile(optimizer='adam', 
           loss='binary_crossentropy', 
           metrics=['accuracy'])

# Train the model with 20 epochs
history = nn.fit(X_train_scaled, 
                 y_train, 
                 epochs=20,  # Reduced number of epochs
                 batch_size=64, 
                 validation_data=(X_test_scaled, y_test), 
                 verbose=2)

Model: "sequential_13"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_64 (Dense)                │ (None, 128)            │         4,992 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_65 (Dense)                │ (None, 64)             │         8,256 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ leaky_re_lu_6 (LeakyReLU)       │ (None, 64)             │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_66 (Dense)                │ (None, 32)             │         2,080 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_67 (Dense)                │ (None, 16)             │           528 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_68 (Dense)                │ (None, 8)              │           136 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_69 (Dense)                │ (None, 1)              │             9 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 16,001 (62.50 KB)

 Trainable params: 16,001 (62.50 KB)

 Non-trainable params: 0 (0.00 B)

Epoch 1/20
429/429 - 4s - 9ms/step - accuracy: 0.7143 - loss: 0.5767 - val_accuracy: 0.7194 - val_loss: 0.5676
Epoch 2/20
429/429 - 1s - 3ms/step - accuracy: 0.7283 - loss: 0.5576 - val_accuracy: 0.7224 - val_loss: 0.5617
Epoch 3/20
429/429 - 1s - 3ms/step - accuracy: 0.7300 - loss: 0.5547 - val_accuracy: 0.7254 - val_loss: 0.5613
Epoch 4/20
429/429 - 1s - 3ms/step - accuracy: 0.7311 - loss: 0.5524 - val_accuracy: 0.7229 - val_loss: 0.5612
Epoch 5/20
429/429 - 1s - 3ms/step - accuracy: 0.7338 - loss: 0.5505 - val_accuracy: 0.7293 - val_loss: 0.5610
Epoch 6/20
429/429 - 1s - 3ms/step - accuracy: 0.7321 - loss: 0.5492 - val_accuracy: 0.7287 - val_loss: 0.5571
Epoch 7/20
429/429 - 1s - 3ms/step - accuracy: 0.7325 - loss: 0.5476 - val_accuracy: 0.7251 - val_loss: 0.5585
Epoch 8/20
429/429 - 2s - 4ms/step - accuracy: 0.7342 - loss: 0.5469 - val_accuracy: 0.7290 - val_loss: 0.5579
Epoch 9/20
429/429 - 1s - 3ms/step - accuracy: 0.7336 - loss: 0.5461 - val_accuracy: 0.7241 - val_loss: 0.5594
E

In [54]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

215/215 - 0s - 2ms/step - accuracy: 0.7262 - loss: 0.5603
Loss: 0.5603225827217102, Accuracy: 0.7262390851974487


- The reduced number of epoch to the training regimen results 72.62% accuracy in the model.
- Despite the previous attempts, the model's accuracy remains at 72%.

In [57]:
# Export our model to HDF5 file
#  YOUR CODE GOES HERE
nn.save("AlphabetSoupCharity_Optimization.h5")